### PR TITLE
[WIP] some modifictions in dict.cpp functions

### PR DESCRIPTION
**Pull request by srajangarg**
Removing few redundancies and other misc changes.

### CODE REVIEW

**Comment by srajangarg** [comment]
@isuruf I was playing around with the `dict.cpp` file. I moved some methods from `dict.h` to `dict.cpp`. And now `make` doesn't complete? It's giving some error during linking `test_parser` for me. Any idea as to what's wrong? The changes are pushed to this PR , and here's the log I'm getting. (You can see it on travis too, after it finishes)
```
[  1%] Building CXX object symengine/CMakeFiles/symengine.dir/basic.cpp.o
[  2%] Building CXX object symengine/CMakeFiles/symengine.dir/dict.cpp.o
[  3%] Building CXX object symengine/CMakeFiles/symengine.dir/symbol.cpp.o
[  4%] Building CXX object symengine/CMakeFiles/symengine.dir/number.cpp.o
[  5%] Building CXX object symengine/CMakeFiles/symengine.dir/integer.cpp.o
[  5%] Building CXX object symengine/CMakeFiles/symengine.dir/rational.cpp.o
[  6%] Building CXX object symengine/CMakeFiles/symengine.dir/complex.cpp.o
[  7%] Building CXX object symengine/CMakeFiles/symengine.dir/constants.cpp.o
[  8%] Building CXX object symengine/CMakeFiles/symengine.dir/add.cpp.o
[  9%] Building CXX object symengine/CMakeFiles/symengine.dir/mul.cpp.o
[ 10%] Building CXX object symengine/CMakeFiles/symengine.dir/pow.cpp.o
[ 11%] Building CXX object symengine/CMakeFiles/symengine.dir/functions.cpp.o
[ 12%] Building CXX object symengine/CMakeFiles/symengine.dir/monomials.cpp.o
[ 12%] Building CXX object symengine/CMakeFiles/symengine.dir/polynomial.cpp.o
[ 13%] Building CXX object symengine/CMakeFiles/symengine.dir/series.cpp.o
[ 14%] Building CXX object symengine/CMakeFiles/symengine.dir/series_generic.cpp.o
[ 15%] Building CXX object symengine/CMakeFiles/symengine.dir/rings.cpp.o
[ 16%] Building CXX object symengine/CMakeFiles/symengine.dir/ntheory.cpp.o
[ 17%] Building CXX object symengine/CMakeFiles/symengine.dir/dense_matrix.cpp.o
[ 18%] Building CXX object symengine/CMakeFiles/symengine.dir/sparse_matrix.cpp.o
[ 18%] Building CXX object symengine/CMakeFiles/symengine.dir/matrix.cpp.o
[ 19%] Building CXX object symengine/CMakeFiles/symengine.dir/visitor.cpp.o
[ 20%] Building CXX object symengine/CMakeFiles/symengine.dir/eval_double.cpp.o
[ 21%] Building CXX object symengine/CMakeFiles/symengine.dir/diophantine.cpp.o
[ 22%] Building CXX object symengine/CMakeFiles/symengine.dir/cwrapper.cpp.o
[ 23%] Building CXX object symengine/CMakeFiles/symengine.dir/printer.cpp.o
[ 24%] Building CXX object symengine/CMakeFiles/symengine.dir/real_double.cpp.o
[ 25%] Building CXX object symengine/CMakeFiles/symengine.dir/complex_double.cpp.o
[ 25%] Building CXX object symengine/CMakeFiles/symengine.dir/expand.cpp.o
[ 26%] Building CXX object symengine/CMakeFiles/symengine.dir/expression.cpp.o
[ 27%] Building CXX object symengine/CMakeFiles/symengine.dir/numer_denom.cpp.o
[ 28%] Building CXX object symengine/CMakeFiles/symengine.dir/derivative.cpp.o
[ 29%] Building CXX object symengine/CMakeFiles/symengine.dir/parser.cpp.o
[ 30%] Building CXX object symengine/CMakeFiles/symengine.dir/sets.cpp.o
[ 31%] Linking CXX static library libsymengine.a
[ 33%] Built target symengine
[ 35%] Built target catch
[ 36%] Linking CXX executable test_rcp
[ 37%] Built target test_rcp
Scanning dependencies of target test_parser
[ 38%] Building CXX object symengine/tests/basic/CMakeFiles/test_parser.dir/test_parser.cpp.o
[ 39%] Linking CXX executable test_parser
../../libsymengine.a(add.cpp.o): In function `SymEngine::Add::__eq__(SymEngine::Basic const&) const':
add.cpp:(.text+0x1f3): undefined reference to `bool SymEngine::umap_eq<std::unordered_map<SymEngine::RCP<SymEngine::Basic const>, SymEngine::RCP<SymEngine::Number const>, SymEngine::RCPBasicHash, SymEngine::RCPBasicKeyEq, std::allocator<std::pair<SymEngine::RCP<SymEngine::Basic const> const, SymEngine::RCP<SymEngine::Number const> > > > >(std::unordered_map<SymEngine::RCP<SymEngine::Basic const>, SymEngine::RCP<SymEngine::Number const>, SymEngine::RCPBasicHash, SymEngine::RCPBasicKeyEq, std::allocator<std::pair<SymEngine::RCP<SymEngine::Basic const> const, SymEngine::RCP<SymEngine::Number const> > > > const&, std::unordered_map<SymEngine::RCP<SymEngine::Basic const>, SymEngine::RCP<SymEngine::Number const>, SymEngine::RCPBasicHash, SymEngine::RCPBasicKeyEq, std::allocator<std::pair<SymEngine::RCP<SymEngine::Basic const> const, SymEngine::RCP<SymEngine::Number const> > > > const&)'
../../libsymengine.a(add.cpp.o): In function `SymEngine::Add::compare(SymEngine::Basic const&) const':
add.cpp:(.text+0x2127): undefined reference to `int SymEngine::map_compare<std::map<SymEngine::RCP<SymEngine::Basic const>, SymEngine::RCP<SymEngine::Number const>, SymEngine::RCPBasicKeyLess, std::allocator<std::pair<SymEngine::RCP<SymEngine::Basic const> const, SymEngine::RCP<SymEngine::Number const> > > > >(std::map<SymEngine::RCP<SymEngine::Basic const>, SymEngine::RCP<SymEngine::Number const>, SymEngine::RCPBasicKeyLess, std::allocator<std::pair<SymEngine::RCP<SymEngine::Basic const> const, SymEngine::RCP<SymEngine::Number const> > > > const&, std::map<SymEngine::RCP<SymEngine::Basic const>, SymEngine::RCP<SymEngine::Number const>, SymEngine::RCPBasicKeyLess, std::allocator<std::pair<SymEngine::RCP<SymEngine::Basic const> const, SymEngine::RCP<SymEngine::Number const> > > > const&)'
```
There are few more similar errors, that I've not pasted.

**Comment by isuruf** [comment]
What a template does is, it generates code for each different type. So, when you are using a templated function, either a compiled function has to be linked in or the implementation has to be there for the compiler to generate code.

When code is moved to `dict.cpp`, only code that can access the compiled object `dict.o` can properly find the generated code or generate new code for the type. So, when external programs like test files try to find the implementation it cannot find them. Library is compiled fine because your implementation is in `dict.cpp` and not in an anonymous namespace. (It the implementations were in an anonymous namespace, then only the functions in `dict.cpp` can find the implementation). In summary, templated functions implementations need to be in the header.

**Comment by srajangarg** [comment]
Thanks! Explanation is sort of clear to me. Does this mean template functions can never be kept in the `.cpp` file? (If some external program want to acess it) Does this also hold for template classes? Also what about member functions of template classes?

**Comment by isuruf** [comment]
Yes to all

**Comment by srajangarg** [comment]
Isn't this very inconvenient? Is there a workaround?

**Comment by isuruf** [comment]
What do you mean?

File: symengine/dict.h [hunk 1]
```diff
@@ -95,19 +123,55 @@ bool map_eq(const T &A, const T &B)
...
+    for (; a != A.end(); ++a, ++b) {
+        cmp = (*a)->__cmp__(**b);
+        if (cmp != 0)
+            return cmp;
+    }
+    return 0;
+}
+
+template <class T>
+int vec_compare(const T &A, const T &B)
```
Comment (isuruf) [inline symengine/dict.h]: `set_eq` and `vec_eq` have identical code right? Use one function for both

**Comment by srajangarg** [comment]
Apart from the `vec_set` methods and the removal of `map_int_Expr` and `map_uint_mpz`, this PR has basically moved around blocks, for clarity.

File: symengine/dict.h [hunk 2]
```diff
@@ -43,6 +43,18 @@ typedef std::map<RCP<const Integer>, unsigned, RCPIntegerKeyLess>
     map_integer_uint;
 typedef std::map<unsigned, integer_class> map_uint_mpz;
 typedef std::map<int, Expression> map_int_Expr;
+//! Part of umap_vec_mpz:
+typedef struct {
+    inline std::size_t operator()(const vec_int &k) const
+    {
+        std::size_t h = 0;
+        for (const auto &p : k) {
+            h = (h << 4) + p;
```
Comment (isuruf) [inline symengine/dict.h]: This is not a very good hash. You should use `hash_combine` here

Comment (srajangarg) [inline symengine/dict.h]: This was being used before, I've just moved it to the top, near the typedefs. Should I change it?

Comment (isuruf) [inline symengine/dict.h]: Right. We'll change it later

**Pull request merged**